### PR TITLE
GH-2401: Deserialize a slice instead of allocating a vec.

### LINKED ISSUE #2401: Do not allocate temporary vector to deserialize data from Global State
We allocate a Vec from data read from lmdb just to consume it later in deserializer. Profiling 50k dictionary writes showed that about 7% probed time was spent in from_vec (which is the deserializing routine we used). After this simple change new flamegraph showed only 0.1% time for the same test run.

### PR DESCRIPTION
**Pull request by mpapierski**
Closes #2401 

This PR avoids allocating temporary Vec to deserialize data from the data returned by `lmdb::Transaction::get.`

### CODE REVIEW

**Review by goral09 — APPROVED** [review]
I support the change made in this PR but let's give @Fraser999 an opportunity to review as well.

File: types/src/bytesrepr.rs [hunk 1]
```diff
@@ -139,6 +139,19 @@ pub fn deserialize<T: FromBytes>(bytes: Vec<u8>) -> Result<T, Error> {
     }
 }
 
+/// Deserializes slice of bbytes into an instance of `T`.
```
Comment (Fraser999) [inline types/src/bytesrepr.rs]:
```suggestion
/// Deserializes a slice of bytes into an instance of `T`.
```

File: types/src/bytesrepr.rs [hunk 2]
```diff
@@ -139,6 +139,19 @@ pub fn deserialize<T: FromBytes>(bytes: Vec<u8>) -> Result<T, Error> {
...
+///
+/// Returns an error if the bytes cannot be deserialized into `T` or if not all of the input bytes
+/// are consumed in the operation.
+pub fn deserialize_slice<T: FromBytes>(bytes: &[u8]) -> Result<T, Error> {
+    let (t, remainder) = T::from_bytes(bytes)?;
```
Comment (Fraser999) [inline types/src/bytesrepr.rs]:
```suggestion
pub fn deserialize_from_slice<I: AsRef<[u8]>, O: FromBytes>(bytes: I) -> Result<O, Error> {
    let (t, remainder) = O::from_bytes(bytes.as_ref())?;
```

**Review by Fraser999 — COMMENTED** [review]
It looks like we could use the new `deserialize_from_slice` function in the following places:
* [InMemoryEnvironment::dump()](https://github.com/casper-network/casper-node/blob/dev/execution_engine/src/storage/trie_store/operations/tests/mod.rs#L1009-L1010)
* [benchmark](https://github.com/casper-network/casper-node/blob/dev/types/benches/bytesrepr_bench.rs#L317) to avoid cloning
* [`CLValue::into_t()`](https://github.com/casper-network/casper-node/blob/dev/types/src/cl_value.rs#L92) to avoid allocating a new vec

File: types/src/bytesrepr.rs [hunk 3]
```diff
@@ -144,6 +144,19 @@ pub fn deserialize<T: FromBytes>(bytes: Vec<u8>) -> Result<T, Error> {
     }
 }
 
+/// Deserializes a slice of bytes into an instance of `T`.
+///
+/// Returns an error if the bytes cannot be deserialized into `T` or if not all of the input bytes
+/// are consumed in the operation.
+pub fn deserialize_from_slice<I: AsRef<[u8]>, O: FromBytes>(bytes: I) -> Result<O, Error> {
```
Comment (Fraser999) [inline types/src/bytesrepr.rs:151]: It'd be good to add this to the `test_serialization_roundtrip` function:
```rust
    let deserialized =
        deserialize_from_slice::<_, T>(&serialized).expect("Unable to deserialize data");
    assert!(*t == deserialized);
```

**Review by Fraser999 — APPROVED** [review]
Optimistic approval given the final comment is addressed.

File: types/src/bytesrepr.rs (hunk 3 above)
Comment (mpapierski) [reply types/src/bytesrepr.rs:151]: Added

**Pull request merged**